pkg/gitops/profile.go

```diff
@@ -0,0 +1,230 @@
+
+	defaultProfileGitHost = "github.com"
+	defaultProfileGitPath = "weaveworks/eks-gitops-example.git"
+)
+
+// GitOptions holds options for cloning a git repository
+type GitOptions struct {
+	URL    string
+	Branch string
+}
+
+// CloneFunc is a function that clones a Git repository
+type CloneFunc func(ctx context.Context, path string, options *git.CloneOptions) error
+
+// Profile represents a GitOps profile
+type Profile struct {
+	ClusterName string
+	Path        string
+	CloneFunc   CloneFunc
+	Fs          afero.Fs
+	IO          afero.Afero
+}
+
+type jsonPatch struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+// ManifestFile represents a manifest file with data
+type ManifestFile struct {
+	Name string
+	Data []byte
+}
+
+// Generate clones the specified Git repo in a base directory and generates overlays if the Git repo
+// points to a profile repo
+func (p *Profile) Generate(ctx context.Context, o GitOptions) error {
+	err := p.CloneFunc(ctx, filepath.Join(p.Path, baseDir), &git.CloneOptions{
+		URL:           o.URL,
+		ReferenceName: plumbing.NewBranchReferenceName(o.Branch),
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "error cloning repository")
+	}
+
+	gitEndpoint, err := transport.NewEndpoint(o.URL)
+	if err != nil {
+		return errors.Wrap(err, "error parsing Git URL")
+	}
+
+	if !matchesDefaultProfile(gitEndpoint) {
+		return nil
+	}
+
+	overlaysPath := filepath.Join(p.Path, overlaysDir)
+
+	if err := p.Fs.MkdirAll(overlaysPath, 0755); err != nil {
+		return errors.Wrapf(err, "error creating overlays dir: %q", overlaysDir)
+	}
+
+	if err := p.writeFluxConfig(); err != nil {
+		return errors.Wrapf(err, "error writing Flux config")
+	}
+
+	manifestFiles, err := p.MakeOverlays()
+	if err != nil {
+		return err
+	}
+
+	for _, manifestFile := range manifestFiles {
+		err := p.IO.WriteFile(filepath.Join(overlaysPath, manifestFile.Name), manifestFile.Data, 0644)
+		if err != nil {
+			return errors.Wrapf(err, "error writing overlay manifest: %q", manifestFile.Name)
+		}
+	}
+
+	return nil
+}
+
+func (p *Profile) writeFluxConfig() error {
+	type generator struct {
+		Command string `json:"command"`
+	}
+	type commandUpdated struct {
+		Generators []generator `json:"generators"`
+	}
+
+	fluxConfig := struct {
+		Version        int            `json:"version"`
+		CommandUpdated commandUpdated `json:"commandUpdated"`
+	}{
+		Version: 1,
+		CommandUpdated: commandUpdated{
+			Generators: []generator{
+				{
+					Command: "kustomize build overlays/cluster-components",
+				},
+			},
+		},
+	}
+
+	data, err := yaml.Marshal(fluxConfig)
+	if err != nil {
+		return errors.Wrap(err, "error marshaling Flux config")
+	}
+
+	return p.IO.WriteFile(filepath.Join(p.Path, ".flux.yaml"), data, 0644)
+}
+
+func matchesDefaultProfile(e *transport.Endpoint) bool {
+	return e.Host == defaultProfileGitHost && ((e.Protocol == "ssh" && e.Path == defaultProfileGitPath) ||
```

Comment:
don't we need to also check that it is a specified revision? Otherwise we will need to make every change in the repo backwards compatible.

Reply:
That is out of scope. Right now, we're assuming a recent version of `eksctl` will work with the latest revision. If we hard-code the revision and check for equality here, we'd have to update the client `eksctl` every time we push a change to the base repo, even if it doesn't necessitate changes to manifest generation. There's also a proposal to off-load manifest generation to an HTTP server that'd avoid the need to keep eksctl up-to-date with the base repo: https://docs.google.com/document/d/1riPEgKQY-Yei0-nG3Vl7zPrK6Wln4iGORUECsrcC50Q/edit#heading=h.uqo7rv6ekotf